### PR TITLE
button widget

### PR DESCRIPTION
**Pull request by willmcgugan**
- Implements a very basic button widget
- Drops DEFAULT_STYLES in favor of a CSS class var
- Changed type of `classes` argument in Widget from a set to a space-separated list, which is often easier to type, especially with the typical case of a single class.
- Adds App.exit which exits the app with a return value 

https://user-images.githubusercontent.com/554369/165757614-a91a690a-1891-45e0-8c56-70a0ca4a4014.mov

### CODE REVIEW

File: sandbox/basic.css [hunk 1]
```diff
@@ -41,26 +41,29 @@ App > Screen {
   background: $primary-darken-2;
   color: $text-primary-darken-2 ;
   border-right: outer $primary-darken-3;
+  content-align: center middle;
```
Comment (willmcgugan) [inline sandbox/basic.css:44]: New rule to align content within a widget.

File: src/textual/app.py [hunk 2]
```diff
@@ -76,8 +78,11 @@ class ActionError(Exception):
     pass
 
 
+ReturnType = TypeVar("ReturnType")
+
+
 @rich.repr.auto
-class App(DOMNode):
+class App(Generic[ReturnType], DOMNode):
```
Comment (willmcgugan) [inline src/textual/app.py:85]: This allows you to specify the return type on `run()` So if you want to return an int you might do this:

```python
class MyApp(App[int]):
```

File: src/textual/color.py [hunk 3]
```diff
@@ -253,7 +258,9 @@ def parse(cls, color_text: str | Color) -> Color:
         """
         if isinstance(color_text, Color):
             return color_text
-        ansi_color = ANSI_COLOR_TO_RGB.get(color_text)
+        if color_text == "transparent":
+            return TRANSPARENT
```
Comment (willmcgugan) [inline src/textual/color.py:262]: Kind of a psuedo-color. Might roll this in to the color names mapping at some point.

File: src/textual/widgets/_button.py [hunk 4]
```diff
@@ -1,67 +1,75 @@
...
+    
+    Button {
+        width: auto;
+        height: 3;
+        padding: 0 2;
+        background: $primary;
+        color: $text-primary;
+        content-align: center middle;
+        border: tall $primary-lighten-3;        
+        margin: 1;
```
Comment (darrenburns) [inline src/textual/widgets/_button.py:27]: IMO Textual widgets should have 0 margin by default, and it's up to the user of the widget if they want to supply margin based on the context they wish to use the widget in.

File: src/textual/widgets/_button.py [hunk 5]
```diff
@@ -1,67 +1,75 @@
...
+        min-width:16;
```
Comment (darrenburns) [inline src/textual/widgets/_button.py:28]: I'd maybe leave this to the user as well?

If I were to drop a button into my app and wanted it to be small e.g. a button with "Yes" or "No", I would need to know that I need to override the "min-width" property which isn't intuitive to me.

It also means I wouldn't be able to set the `width` property to anything below 16, right? It seems `min-width` takes precedence.

**Review by darrenburns — CHANGES_REQUESTED** [review]
Some comments, questions, suggestions. Also this PR seems to break hot reloading somehow - on making a CSS change on any sandbox app my screen just goes black.

File: src/textual/app.py (hunk 2 above)
Comment (willmcgugan) [reply src/textual/app.py:85]: Came from a Textual issue I can't seem to find now. But a user wan't to be able to launch TUIs within a larger app to get info from the user. The data could be an int, str, or something more complex.

File: src/textual/widgets/_button.py (hunk 5 above)
Comment (willmcgugan) [reply src/textual/widgets/_button.py:28]: Done. Although without a min-width, the buttons may be all different sizes which may not look good.

File: src/textual/widgets/_button.py (hunk 4 above)
Comment (willmcgugan) [reply src/textual/widgets/_button.py:27]: Without margin on the vertical, the buttons bump up against each other on the vertical axis, which doesn't look good. But you don't get that on the horizontal axis. So I've set margin to 1 0.

This CSS is a starting point though. Plenty of tweaking to do.

File: sandbox/basic.css (hunk 1 above)
Comment (willmcgugan) [reply sandbox/basic.css:44]: align moves the Widget, content-align moves the text inside a widget. The docs will have nice diagrams to show the difference.

I would expect there to be little reason to adjust content-align for a pre-built widgets. For instance a button will likely always have content-align center middle.

Comment (DrBenton) [reply sandbox/basic.css:44]: It's very subjective, and I'm probably bike-shedding :sweat_smile: , but... wouldn't something like `justify`, `text-justify`, `justify-items` or `justify-content` be more differentiated to `align` than `content-align`? (they all are existing CSS property names)

**Pull request merged**